File: solutions/cpp/kindergarten-garden/3/kindergarten_garden.cpp
#include "kindergarten_garden.h"
#include <algorithm>
#include <iterator>
#include <cassert>
#include <map>
#include <stdexcept>
#include <vector>
namespace kindergarten_garden {

// TODO: add your solution here

    const std::vector<std::string> STUDENTS{
        "Alice",
        "Bob",
        "Charlie",
        "David",
        "Eve",
        "Fred",
        "Ginny",
        "Harriet",
        "Ileana",
        "Joseph",
        "Kincaid",
        "Larry"
    };

    std::array<Plants, 4> plants(const std::string& plant_assignment, const std::string& student) {
        const size_t row_break = plant_assignment.find("\n");
        assert(row_break != std::string::npos);
        const size_t row2_start = row_break + 1;
        std::string row[2] = {
            plant_assignment.substr(0, row_break),
            plant_assignment.substr(row_break + 1)
        };
        auto student_index = static_cast<size_t>(std::distance(STUDENTS.begin(), std::find(STUDENTS.begin(), STUDENTS.end(), student)));
        if (student_index == STUDENTS.size())
            throw std::domain_error("unknown student");
        
        auto plant_index = student_index * 2;
        if (plant_index >= row_break)
            throw std::domain_error("invalid argument value");

        return {
            static_cast<Plants>(plant_assignment[plant_index]), 
            static_cast<Plants>(plant_assignment[plant_index + 1]),
            static_cast<Plants>(plant_assignment[row2_start + plant_index]), 
            static_cast<Plants>(plant_assignment[row2_start + plant_index + 1])
        };
    }

}  // namespace kindergarten_garden

File: solutions/python/kindergarten-garden/1/kindergarten_garden.py
class Garden:
    Plants = {"G": "Grass", "C": "Clover", "R": "Radishes", "V": "Violets"}

    def __init__(self, diagram: str, students: list[str] = None):
        self.diagram = diagram
        if students is None:
            students = ["Alice", "Bob", "Charlie", "David", "Eve", "Fred",
                        "Ginny", "Harriet", "Ileana", "Joseph", "Kincaid", "Larry"]
        self.students = sorted(students)
        student_count = len(self.students)
        self.diagram: list[str] = diagram.split()
        diagram_len = list(map(len, self.diagram))
        if not all(map(lambda x: x == diagram_len[0], diagram_len)):
            raise ValueError("Invalid diagram format")

    def plants(self, student: str) -> list[str]:
        if student not in self.students:
            raise ValueError(f"'{student}' not found")
        index = self.students.index(student) * 2
        return [Garden.Plants.get(self.diagram[0][index], "unknown"),
                Garden.Plants.get(self.diagram[0][index + 1], "unknown"),
                Garden.Plants.get(self.diagram[1][index], "unknown"),
                Garden.Plants.get(self.diagram[1][index + 1], "unknown")]

